<a href="https://colab.research.google.com/github/aimlalchemists/personal-loan-campaign-modelling/blob/main/AllLife_Bank_Personal_Loan_Campign_Modelling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<span style="font-family: Helvetica; font-weight:bold;font-size:2.5em;color:navy">Machine Learning: AllLife Bank Personal Loan Campaign</span>

<span style="font-family: Helvetica; font-weight:bold;font-size:1.5em;color:green">Muthuraman Muthusubramanian</span>




<span style="font-family: Helvetica; font-weight:bold;font-size:2.5em;color:navy">Problem Statement</span>

### Context

AllLife Bank is a US bank that has a growing customer base. The majority of these customers are liability customers (depositors) with varying sizes of deposits. The number of customers who are also borrowers (asset customers) is quite small, and the bank is interested in expanding this base rapidly to bring in more loan business and in the process, earn more through the interest on loans. In particular, the management wants to explore ways of converting its liability customers to personal loan customers (while retaining them as depositors).

A campaign that the bank ran last year for liability customers showed a healthy conversion rate of over 9% success. This has encouraged the retail marketing department to devise campaigns with better target marketing to increase the success ratio.

You as a Data scientist at AllLife bank have to build a model that will help the marketing department to identify the potential customers who have a higher probability of purchasing the loan.

### Objective

To predict whether a liability customer will buy personal loans, to understand which customer attributes are most significant in driving purchases, and identify which segment of customers to target more.

### Data Dictionary
* `ID`: Customer ID
* `Age`: Customer’s age in completed years
* `Experience`: #years of professional experience
* `Income`: Annual income of the customer (in thousand dollars)
* `ZIP Code`: Home Address ZIP code.
* `Family`: the Family size of the customer
* `CCAvg`: Average spending on credit cards per month (in thousand dollars)
* `Education`: Education Level. 1: Undergrad; 2: Graduate;3: Advanced/Professional
* `Mortgage`: Value of house mortgage if any. (in thousand dollars)
* `Personal_Loan`: Did this customer accept the personal loan offered in the last campaign? (0: No, 1: Yes)
* `Securities_Account`: Does the customer have securities account with the bank? (0: No, 1: Yes)
* `CD_Account`: Does the customer have a certificate of deposit (CD) account with the bank? (0: No, 1: Yes)
* `Online`: Do customers use internet banking facilities? (0: No, 1: Yes)
* `CreditCard`: Does the customer use a credit card issued by any other Bank (excluding All life Bank)? (0: No, 1: Yes)

In [ ]:
# verify the version numbers of packages installed.
!pip freeze | grep numpy
!pip freeze | grep pandas
!pip freeze | grep matplotlib
!pip freeze | grep seaborn
!pip freeze | grep scikit-learn
!pip freeze | grep scipy

numpy==1.25.2
geopandas==0.13.2
pandas==2.0.3
pandas-datareader==0.10.0
pandas-gbq==0.19.2
pandas-stubs==2.0.3.230814
sklearn-pandas==2.2.0
matplotlib==3.7.1
matplotlib-inline==0.1.7
matplotlib-venn==0.11.10
seaborn==0.13.1
scikit-learn==1.2.2
scipy==1.11.4


In [ ]:
# install necessary libraries.
!pip install numpy pandas matplotlib seaborn scikit-learn sklearn-pandas -q --user

**Note**: *After running the above cell, kindly restart the notebook kernel and run all cells sequentially from the start again.*

<span style="font-family: Helvetica; font-weight:bold;font-size:2.5em;color:navy">Importing necessary libraries</span>

In [1]:
# Import libraries to read and manipulate the data
import pandas as pd
import numpy as np
# Import plotting libraries for visuals and statistic analysis.
import math
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
# Import libraries for model predictions and decision trees.
from sklearn import metrics, tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, GridSearchCV

# To get different metrics scores
from sklearn.metrics import (confusion_matrix, classification_report,
                             accuracy_score, precision_score, recall_score, f1_score)

# To ignore the warnings.
import warnings
warnings.filterwarnings("ignore")

# Import drive libraries to get data from google drive.
from google.colab import drive
drive.mount('/content/drive')

# Allows plots created with Matplotlib to be displayed directly in the notebook.
%matplotlib inline

# Sets the default parameters for plots.
sns.set()

Mounted at /content/drive


<span style="font-family: Helvetica; font-weight:bold;font-size:2.5em;color:navy">Loading the dataset</span>

In [2]:
# read the data.
data = pd.read_csv("https://raw.githubusercontent.com/aimlalchemists/personal-loan-campaign-modelling/main/Loan_Modelling.csv")

# copy the data into df
df = data.copy()

# count of records.
print(f"There is {df.shape[0]} rows and {df.shape[1]} columns in this dataset.")

There is 5000 rows and 14 columns in this dataset.


In [3]:
# check first 5 rows of data
df.head()

ID  Age  Experience  Income  ZIPCode  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49    91107       4    1.6          1         0   
1   2   45          19      34    90089       3    1.5          1         0   
2   3   39          15      11    94720       1    1.0          1         0   
3   4   35           9     100    94112       1    2.7          2         0   
4   5   35           8      45    91330       4    1.0          2         0   

   Personal_Loan  Securities_Account  CD_Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [4]:
# get the last 5 rows of data.
df.tail()

ID  Age  Experience  Income  ZIPCode  Family  CCAvg  Education  \
4995  4996   29           3      40    92697       1    1.9          3   
4996  4997   30           4      15    92037       4    0.4          1   
4997  4998   63          39      24    93023       2    0.3          3   
4998  4999   65          40      49    90034       3    0.5          2   
4999  5000   28           4      83    92612       3    0.8          1   

      Mortgage  Personal_Loan  Securities_Account  CD_Account  Online  \
4995         0              0                   0           0       1   
4996        85              0                   0           0       1   
4997         0              0                   0           0       0   
4998         0              0                   0           0       1   
4999         0              0                   0           0       1   

      CreditCard  
4995           0  
4996           0  
4997           0  
4998           0  
4999           1

<span style="font-family: Helvetica; font-weight:bold;font-size:2.5em;color:navy">Data Overview</span>

In [5]:
# Understand the not null, count and data type of the 14 columns.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIPCode             5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal_Loan       5000 non-null   int64  
 10  Securities_Account  5000 non-null   int64  
 11  CD_Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


### observations

- 13 columns int64 and one column float64.
- Personal_Loan is the dependent variable.
- Totally 5000 entries
- The Education and Family columns are currently an INT64 data type. This might not be suitable as those are often categorical, not numerical. Need to consider revising the data type for better data representation

In [6]:
# Describe the data
df.describe()

ID          Age   Experience       Income       ZIPCode  \
count  5000.000000  5000.000000  5000.000000  5000.000000   5000.000000   
mean   2500.500000    45.338400    20.104600    73.774200  93169.257000   
std    1443.520003    11.463166    11.467954    46.033729   1759.455086   
min       1.000000    23.000000    -3.000000     8.000000  90005.000000   
25%    1250.750000    35.000000    10.000000    39.000000  91911.000000   
50%    2500.500000    45.000000    20.000000    64.000000  93437.000000   
75%    3750.250000    55.000000    30.000000    98.000000  94608.000000   
max    5000.000000    67.000000    43.000000   224.000000  96651.000000   

            Family        CCAvg    Education     Mortgage  Personal_Loan  \
count  5000.000000  5000.000000  5000.000000  5000.000000    5000.000000   
mean      2.396400     1.937938     1.881000    56.498800       0.096000   
std       1.147663     1.747659     0.839869   101.713802       0.294621   
min       1.000000     0.000000     1.000000     0.000000       0.000000   
25%       1.000000     0.700000     1.000000     0.000000       0.000000   
50%       2.000000     1.500000     2.000000     0.000000       0.000000   
75%       3.000000     2.500000     3.000000   101.000000       0.000000   
max       4.000000    10.000000     3.000000   635.000000       1.000000   

       Securities_Account  CD_Account       Online   CreditCard  
count         5000.000000  5000.00000  5000.000000  5000.000000  
mean             0.104400     0.06040     0.596800     0.294000  
std              0.305809     0.23825     0.490589     0.455637  
min              0.000000     0.00000     0.000000     0.000000  
25%              0.000000     0.00000     0.000000     0.000000  
50%              0.000000     0.00000     1.000000     0.000000  
75%              0.000000     0.00000     1.000000     1.000000  
max              1.000000     1.00000     1.000000     1.000000

In [7]:
# convert the Family and Education as Category column.
df.Family = df.Family.astype("category")
df.Education = df.Education.astype("category")

In [8]:
df.describe().T

count          mean          std      min       25%  \
ID                  5000.0   2500.500000  1443.520003      1.0   1250.75   
Age                 5000.0     45.338400    11.463166     23.0     35.00   
Experience          5000.0     20.104600    11.467954     -3.0     10.00   
Income              5000.0     73.774200    46.033729      8.0     39.00   
ZIPCode             5000.0  93169.257000  1759.455086  90005.0  91911.00   
CCAvg               5000.0      1.937938     1.747659      0.0      0.70   
Mortgage            5000.0     56.498800   101.713802      0.0      0.00   
Personal_Loan       5000.0      0.096000     0.294621      0.0      0.00   
Securities_Account  5000.0      0.104400     0.305809      0.0      0.00   
CD_Account          5000.0      0.060400     0.238250      0.0      0.00   
Online              5000.0      0.596800     0.490589      0.0      0.00   
CreditCard          5000.0      0.294000     0.455637      0.0      0.00   

                        50%       75%      max  
ID                   2500.5   3750.25   5000.0  
Age                    45.0     55.00     67.0  
Experience             20.0     30.00     43.0  
Income                 64.0     98.00    224.0  
ZIPCode             93437.0  94608.00  96651.0  
CCAvg                   1.5      2.50     10.0  
Mortgage                0.0    101.00    635.0  
Personal_Loan           0.0      0.00      1.0  
Securities_Account      0.0      0.00      1.0  
CD_Account              0.0      0.00      1.0  
Online                  1.0      1.00      1.0  
CreditCard              0.0      1.00      1.0

### Observations
- Looks like there are no missing values as per the count column. Still need to verify and confirm if there are any inappropriate values present.
- ID column may not be required if its just a sequence of numbers.
- Age values stand between 23 and 67, with an average of 45.
- Experience shows -3 as minimum, sounds incorrect.
- Not able to get a clarity on CreditCard column. If that's a boolean column, then better to change the name as HasCreditCard or IsCreditCardHolder.

In [9]:
# check for duplicates
df.duplicated().sum()

0

In [10]:
# check for na values
df.isna().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIPCode               0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal_Loan         0
Securities_Account    0
CD_Account            0
Online                0
CreditCard            0
dtype: int64

### Observations
- No duplicates or na values

In [11]:
# Drop the ID column as that's not giving any value add in the data.
df.drop("ID", axis=1, inplace=True)

In [12]:
df.describe(include='all').T

count  unique  top    freq          mean          std  \
Age                 5000.0     NaN  NaN     NaN     45.338400    11.463166   
Experience          5000.0     NaN  NaN     NaN     20.104600    11.467954   
Income              5000.0     NaN  NaN     NaN     73.774200    46.033729   
ZIPCode             5000.0     NaN  NaN     NaN  93169.257000  1759.455086   
Family              5000.0     4.0  1.0  1472.0           NaN          NaN   
CCAvg               5000.0     NaN  NaN     NaN      1.937938     1.747659   
Education           5000.0     3.0  1.0  2096.0           NaN          NaN   
Mortgage            5000.0     NaN  NaN     NaN     56.498800   101.713802   
Personal_Loan       5000.0     NaN  NaN     NaN      0.096000     0.294621   
Securities_Account  5000.0     NaN  NaN     NaN      0.104400     0.305809   
CD_Account          5000.0     NaN  NaN     NaN      0.060400     0.238250   
Online              5000.0     NaN  NaN     NaN      0.596800     0.490589   
CreditCard          5000.0     NaN  NaN     NaN      0.294000     0.455637   

                        min      25%      50%      75%      max  
Age                    23.0     35.0     45.0     55.0     67.0  
Experience             -3.0     10.0     20.0     30.0     43.0  
Income                  8.0     39.0     64.0     98.0    224.0  
ZIPCode             90005.0  91911.0  93437.0  94608.0  96651.0  
Family                  NaN      NaN      NaN      NaN      NaN  
CCAvg                   0.0      0.7      1.5      2.5     10.0  
Education               NaN      NaN      NaN      NaN      NaN  
Mortgage                0.0      0.0      0.0    101.0    635.0  
Personal_Loan           0.0      0.0      0.0      0.0      1.0  
Securities_Account      0.0      0.0      0.0      0.0      1.0  
CD_Account              0.0      0.0      0.0      0.0      1.0  
Online                  0.0      0.0      1.0      1.0      1.0  
CreditCard              0.0      0.0      0.0      1.0      1.0

### Observatinos
- There are 4 unique values for Family and 3 unique values for Education.
- There are only two unique values in Personal_Loan, Securities_Account, cd_Account, online and CreditCard columns.
- Age column has an average value of 45 with a standard deviation of approximately 11.4. The ages range from 23 to 67.
- Experience column has an average of 20 years with a standard deviation of 11.5 years. The experience ranges from -3 to 43 years. We will further examine the negative value.
- Income column has an average of 74K with a standard deviation of 46K. The income values span from 8K to 224K.
- Ccavg column has an average of 1.93 with a standard deviation of 1.7. The values in this column range from 0.0 to 10.0.
- Mortgage column has an average of 56.5K with a standard deviation of 101K. The standard deviation is larger than the mean, which warrants further investigation.
- There are zero values present in the mortgage column, which we will also inspect

## Exploratory Data Analysis.

- EDA is an important part of any project involving data.
- It is important to investigate and understand the data better before building a model with it.
- A few questions have been mentioned below which will help you approach the analysis in the right manner and generate insights from the data.
- A thorough analysis of the data, in addition to the questions mentioned below, should be done.

**Questions**:

1. What is the distribution of mortgage attribute? Are there any noticeable patterns or outliers in the distribution?
2. How many customers have credit cards?
3. What are the attributes that have a strong correlation with the target attribute (personal loan)?
4. How does a customer's interest in purchasing a loan vary with their age?
5. How does a customer's interest in purchasing a loan vary with their education?

## Data Preprocessing

* Missing value treatment
* Feature engineering (if needed)
* Outlier detection and treatment (if needed)
* Preparing data for modeling
* Any other preprocessing steps (if needed)

## Model Building

### Model Evaluation Criterion

*


### Model Building

### Model Performance Improvement

## Model Comparison and Final Model Selection

## Actionable Insights and Business Recommendations


* What recommedations would you suggest to the bank?

___